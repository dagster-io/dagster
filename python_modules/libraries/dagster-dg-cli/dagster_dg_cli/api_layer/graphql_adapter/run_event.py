"""GraphQL adapter for run events."""

from typing import Any, Optional

from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

# Exact GraphQL query from specification
RUN_EVENTS_QUERY = """
query CliRunEventsQuery($runId: ID!, $limit: Int, $afterCursor: String) {
    logsForRun(runId: $runId, limit: $limit, afterCursor: $afterCursor) {
        __typename
        ... on EventConnection {
            events {
                __typename
                ... on MessageEvent {
                    runId
                    message
                    timestamp
                    level
                    stepKey
                    eventType
                }
            }
            cursor
            hasMore
        }
        ... on PythonError {
            message
            stack
        }
        ... on RunNotFoundError {
            message
        }
    }
}
"""


def _filter_events_by_type(events: list[dict], event_type: Optional[str]) -> list[dict]:
    """Client-side filtering logic for event types."""
    if not event_type:
        return events

    # Split comma-separated types and normalize to uppercase
    types = [t.strip().upper() for t in event_type.split(",")]

    filtered = []
    for event in events:
        event_type_val = event.get("eventType")
        if event_type_val and event_type_val.upper() in types:
            filtered.append(event)

    return filtered


def _filter_events_by_step(events: list[dict], step_key: Optional[str]) -> list[dict]:
    """Client-side filtering logic for step keys."""
    if not step_key:
        return events

    filtered = []
    for event in events:
        event_step = event.get("stepKey", "")
        if event_step and step_key.lower() in event_step.lower():
            filtered.append(event)

    return filtered


def get_run_events_via_graphql(
    client: IGraphQLClient,
    run_id: str,
    limit: int = 100,
    after_cursor: Optional[str] = None,
    event_type: Optional[str] = None,
    step_key: Optional[str] = None,
) -> dict[str, Any]:
    """Get run events via GraphQL with client-side filtering."""
    variables = {"runId": run_id, "limit": limit}
    if after_cursor:
        variables["afterCursor"] = after_cursor

    result = client.execute(RUN_EVENTS_QUERY, variables)

    logs_result = result.get("logsForRun")
    if not logs_result:
        raise Exception("Empty response from GraphQL API")

    typename = logs_result.get("__typename")

    if typename != "EventConnection":
        error_msg = logs_result.get("message", f"Unknown error: {typename}")
        raise Exception(error_msg)

    # Extract and filter events
    events_data = logs_result.get("events", [])

    # Apply client-side filters
    if event_type:
        events_data = _filter_events_by_type(events_data, event_type)

    if step_key:
        events_data = _filter_events_by_step(events_data, step_key)

    return {
        "events": events_data,
        "cursor": logs_result.get("cursor"),
        "hasMore": logs_result.get("hasMore", False),
    }
