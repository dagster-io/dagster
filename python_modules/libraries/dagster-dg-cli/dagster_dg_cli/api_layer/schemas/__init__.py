"""Pydantic schemas for Dagster Plus API."""
