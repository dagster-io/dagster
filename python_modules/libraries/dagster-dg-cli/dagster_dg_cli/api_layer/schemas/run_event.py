"""Run event schema definitions."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class RunEventLevel(str, Enum):
    """Event severity levels."""

    CRITICAL = "CRITICAL"
    ERROR = "ERROR"
    WARNING = "WARNING"
    INFO = "INFO"
    DEBUG = "DEBUG"


class RunEvent(BaseModel):
    """Single run event model."""

    run_id: str
    message: str
    timestamp: str  # ISO 8601 timestamp
    level: RunEventLevel
    step_key: Optional[str] = None
    event_type: str

    class Config:
        from_attributes = True


class RunEventList(BaseModel):
    """Paginated run events response."""

    items: list[RunEvent]
    total: int
    cursor: Optional[str] = None
    has_more: bool = False

    class Config:
        from_attributes = True
