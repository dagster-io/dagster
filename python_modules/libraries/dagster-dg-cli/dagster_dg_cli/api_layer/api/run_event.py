"""Run events API implementation."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.run_event import get_run_events_via_graphql
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.run_event import RunEventList


@dataclass(frozen=True)
class DgApiRunEventApi:
    """API for run events operations."""

    client: IGraphQLClient

    def get_events(
        self,
        run_id: str,
        event_type: Optional[str] = None,
        step_key: Optional[str] = None,
        limit: int = 100,
        after_cursor: Optional[str] = None,
    ) -> "RunEventList":
        """Get run events with filtering options."""
        from dagster_dg_cli.api_layer.schemas.run_event import RunEvent, RunEventLevel, RunEventList

        events_data = get_run_events_via_graphql(
            self.client,
            run_id=run_id,
            limit=limit,
            after_cursor=after_cursor,
            event_type=event_type,
            step_key=step_key,
        )

        # Convert to Pydantic models
        events = [
            RunEvent(
                run_id=e["runId"],
                message=e["message"],
                timestamp=e["timestamp"],
                level=RunEventLevel[e["level"]],
                step_key=e.get("stepKey"),
                event_type=e["eventType"],
            )
            for e in events_data["events"]
        ]

        return RunEventList(
            items=events,
            total=len(events),
            cursor=events_data.get("cursor"),
            has_more=events_data.get("hasMore", False),
        )
