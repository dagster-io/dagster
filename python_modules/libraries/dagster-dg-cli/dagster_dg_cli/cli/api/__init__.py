"""API commands for interacting with Dagster Plus REST-like interface."""

from dagster_dg_cli.cli.api.cli_group import api_group

__all__ = ["api_group"]
