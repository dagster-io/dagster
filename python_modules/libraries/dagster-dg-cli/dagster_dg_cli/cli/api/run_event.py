"""Run events API commands following GitHub CLI patterns."""

import json

import click
from dagster_dg_core.utils import DgClickCommand, DgClickGroup
from dagster_dg_core.utils.telemetry import cli_telemetry_wrapper
from dagster_shared.plus.config import DagsterPlusCliConfig
from dagster_shared.plus.config_utils import dg_api_options

from dagster_dg_cli.api_layer.api.run_event import DgApiRunEventApi
from dagster_dg_cli.cli.api.client import create_dg_api_graphql_client


def format_run_events_table(events, run_id: str) -> str:
    """Format run events as human-readable table."""
    if not events.items:
        return f"No events found for run {run_id}"

    lines = [f"Events for run {run_id}:", ""]

    # Table header
    header = f"{'TIMESTAMP':<20} {'TYPE':<20} {'STEP_KEY':<30} {'MESSAGE':<50}"
    separator = "-" * len(header)
    lines.extend([header, separator])

    # Table rows
    for event in events.items:
        timestamp = event.timestamp[:19]  # Truncate to YYYY-MM-DDTHH:MM:SS
        event_type = event.event_type
        step_key = event.step_key or ""
        message = event.message[:47] + "..." if len(event.message) > 50 else event.message

        row = f"{timestamp:<20} {event_type:<20} {step_key:<30} {message:<50}"
        lines.append(row)

    lines.extend(["", f"Total events: {events.total}"])
    return "\n".join(lines)


def format_run_events_json(events, run_id: str) -> str:
    """Format run events as JSON."""
    return json.dumps(
        {
            "run_id": run_id,
            "events": [
                {
                    "runId": event.run_id,
                    "message": event.message,
                    "timestamp": event.timestamp,
                    "level": event.level,
                    "stepKey": event.step_key,
                    "eventType": event.event_type,
                }
                for event in events.items
            ],
            "count": events.total,
        },
        indent=2,
    )


@click.command(name="get", cls=DgClickCommand, unlaunched=True)
@click.argument("run_id", type=str)
@click.option(
    "--type",
    "event_type",
    help="Filter by event type (comma-separated)",
)
@click.option(
    "--step",
    "step_key",
    help="Filter by step key (partial matching)",
)
@click.option(
    "--limit",
    type=int,
    default=100,
    help="Maximum number of events to return",
)
@click.option(
    "--json",
    "output_json",
    is_flag=True,
    help="Output in JSON format for machine readability",
)
@dg_api_options(deployment_scoped=True)
@cli_telemetry_wrapper
@click.pass_context
def get_run_events_command(
    ctx: click.Context,
    run_id: str,
    event_type: str,
    step_key: str,
    limit: int,
    output_json: bool,
    organization: str,
    deployment: str,
    api_token: str,
) -> None:
    """Get run events with filtering options."""
    config = DagsterPlusCliConfig.create_for_deployment(
        deployment=deployment,
        organization=organization,
        user_token=api_token,
    )
    client = create_dg_api_graphql_client(ctx, config)
    api = DgApiRunEventApi(client)

    try:
        events = api.get_events(
            run_id=run_id, event_type=event_type, step_key=step_key, limit=limit
        )

        if output_json:
            output = format_run_events_json(events, run_id)
        else:
            output = format_run_events_table(events, run_id)

        click.echo(output)
    except Exception as e:
        if output_json:
            error_response = {"error": str(e)}
            click.echo(json.dumps(error_response), err=True)
        else:
            click.echo(f"Error querying Dagster Plus API: {e}", err=True)


@click.group(
    name="run-events",
    cls=DgClickGroup,
    unlaunched=True,
    commands={
        "get": get_run_events_command,
    },
)
def run_events_group():
    """Manage run events in Dagster Plus."""
