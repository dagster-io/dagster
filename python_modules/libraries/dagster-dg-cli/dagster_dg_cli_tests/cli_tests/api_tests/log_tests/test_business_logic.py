"""Test log business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

from dagster_dg_cli.api_layer.schemas.run_event import (
    DgApiErrorInfo,
    DgApiRunEvent,
    RunEventLevel,
    RunEventList,
)
from dagster_dg_cli.cli.api.log import format_logs_json, format_logs_table


class TestFormatLogs:
    """Test the log formatting functions."""

    def _create_sample_log_list(self):
        """Create sample RunEventList for testing."""
        events = [
            DgApiRunEvent(
                run_id="test-run-12345",
                message='Starting execution of run for "test_pipeline".',
                timestamp="1641046800000",  # 2022-01-01 14:20:00 UTC (as milliseconds)
                level=RunEventLevel.INFO,
                step_key=None,
                event_type="RunStartEvent",
                error=None,
            ),
            DgApiRunEvent(
                run_id="test-run-12345",
                message='Started execution of step "process_data".',
                timestamp="1641046805000",  # 2022-01-01 14:20:05 UTC
                level=RunEventLevel.DEBUG,
                step_key="process_data",
                event_type="ExecutionStepStartEvent",
                error=None,
            ),
            DgApiRunEvent(
                run_id="test-run-12345",
                message="Loading input from path: /tmp/input.json",
                timestamp="1641046810000",  # 2022-01-01 14:20:10 UTC
                level=RunEventLevel.DEBUG,
                step_key="process_data",
                event_type="MessageEvent",
                error=None,
            ),
            DgApiRunEvent(
                run_id="test-run-12345",
                message='Execution of step "process_data" failed.',
                timestamp="1641046815000",  # 2022-01-01 14:20:15 UTC
                level=RunEventLevel.ERROR,
                step_key="process_data",
                event_type="ExecutionStepFailureEvent",
                error=DgApiErrorInfo(
                    message="ValueError: Invalid input data format\n",
                    className="ValueError",
                    stack=[
                        '  File "/app/pipeline.py", line 42, in process_data\n    data = json.loads(input_str)\n',
                        '  File "/usr/lib/python3.12/json/__init__.py", line 346, in loads\n    return _default_decoder.decode(s)\n',
                        '  File "/usr/lib/python3.12/json/decoder.py", line 337, in decode\n    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n',
                    ],
                    cause=None,
                ),
            ),
            DgApiRunEvent(
                run_id="test-run-12345",
                message="Execution of run for \"test_pipeline\" failed. Steps failed: ['process_data'].",
                timestamp="1641046820000",  # 2022-01-01 14:20:20 UTC
                level=RunEventLevel.ERROR,
                step_key=None,
                event_type="RunFailureEvent",
                error=None,
            ),
        ]
        return RunEventList(items=events, total=5, cursor=None, has_more=False)

    def _create_empty_log_list(self):
        """Create empty RunEventList for testing."""
        return RunEventList(items=[], total=0, cursor=None, has_more=False)

    def _create_log_with_nested_error(self):
        """Create log with nested error causes."""
        return DgApiRunEvent(
            run_id="nested-error-run",
            message="Database connection failed with retry exhausted.",
            timestamp="1641046825000",
            level=RunEventLevel.ERROR,
            step_key="database_query",
            event_type="ExecutionStepFailureEvent",
            error=DgApiErrorInfo(
                message="RetryRequestedFromPolicy: Exceeded max_retries of 3\n",
                className="RetryRequestedFromPolicy",
                stack=[
                    '  File "/app/database.py", line 25, in execute_query\n    return self._execute_with_retry(query)\n',
                    '  File "/app/database.py", line 45, in _execute_with_retry\n    raise RetryRequestedFromPolicy(f"Exceeded max_retries of {max_retries}")\n',
                ],
                cause=DgApiErrorInfo(
                    message="ConnectionError: [Errno 111] Connection refused\n",
                    className="ConnectionError",
                    stack=[
                        '  File "/app/database.py", line 35, in _execute_with_retry\n    result = self.connection.execute(query)\n',
                        '  File "/usr/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute\n    return connection.execute(statement, parameters)\n',
                        '  File "/usr/lib/python3.12/site-packages/psycopg2/__init__.py", line 122, in connect\n    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)\n',
                    ],
                    cause=None,
                ),
            ),
        )

    def _create_log_with_very_long_step_key(self):
        """Create log with very long step key to test truncation."""
        return DgApiRunEvent(
            run_id="truncation-test-run",
            message="Processing large dataset chunk.",
            timestamp="1641046830000",
            level=RunEventLevel.INFO,
            step_key="very_long_step_name_that_exceeds_the_display_limit_and_should_be_truncated",
            event_type="MessageEvent",
            error=None,
        )

    def test_format_logs_table_output(self, snapshot):
        """Test formatting logs as text table."""
        from dagster_shared.utils.timing import fixed_timezone

        log_list = self._create_sample_log_list()
        with fixed_timezone("UTC"):
            result = format_logs_table(log_list, "test-run-12345")

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_logs_json_output(self, snapshot):
        """Test formatting logs as JSON."""
        log_list = self._create_sample_log_list()
        result = format_logs_json(log_list, "test-run-12345")

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_empty_logs_table_output(self, snapshot):
        """Test formatting empty log list as text."""
        log_list = self._create_empty_log_list()
        result = format_logs_table(log_list, "empty-run-67890")

        snapshot.assert_match(result)

    def test_format_empty_logs_json_output(self, snapshot):
        """Test formatting empty log list as JSON."""
        log_list = self._create_empty_log_list()
        result = format_logs_json(log_list, "empty-run-67890")

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_logs_with_nested_errors(self, snapshot):
        """Test formatting logs with nested error causes."""
        from dagster_shared.utils.timing import fixed_timezone

        log_with_nested_error = self._create_log_with_nested_error()
        log_list = RunEventList(items=[log_with_nested_error], total=1)

        with fixed_timezone("UTC"):
            result = format_logs_table(log_list, "nested-error-run")

        snapshot.assert_match(result)

    def test_format_logs_with_nested_errors_json(self, snapshot):
        """Test formatting logs with nested errors as JSON."""
        log_with_nested_error = self._create_log_with_nested_error()
        log_list = RunEventList(items=[log_with_nested_error], total=1)

        result = format_logs_json(log_list, "nested-error-run")

        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_logs_with_long_step_key(self, snapshot):
        """Test formatting logs with step key truncation."""
        from dagster_shared.utils.timing import fixed_timezone

        log_with_long_key = self._create_log_with_very_long_step_key()
        log_list = RunEventList(items=[log_with_long_key], total=1)

        with fixed_timezone("UTC"):
            result = format_logs_table(log_list, "truncation-test-run")

        snapshot.assert_match(result)

    def test_format_logs_with_pagination_info(self, snapshot):
        """Test formatting logs with pagination indicators."""
        from dagster_shared.utils.timing import fixed_timezone

        # Create log list that has more data available
        log_list = RunEventList(
            items=[
                DgApiRunEvent(
                    run_id="paginated-run",
                    message="First log entry",
                    timestamp="1641046800000",
                    level=RunEventLevel.INFO,
                    step_key=None,
                    event_type="MessageEvent",
                    error=None,
                ),
                DgApiRunEvent(
                    run_id="paginated-run",
                    message="Second log entry",
                    timestamp="1641046805000",
                    level=RunEventLevel.DEBUG,
                    step_key="step_1",
                    event_type="MessageEvent",
                    error=None,
                ),
            ],
            total=2,
            cursor="cursor_token_12345",
            has_more=True,
        )

        with fixed_timezone("UTC"):
            result = format_logs_table(log_list, "paginated-run")

        snapshot.assert_match(result)

    def test_format_logs_all_levels(self, snapshot):
        """Test formatting logs with all log levels."""
        from dagster_shared.utils.timing import fixed_timezone

        events = [
            DgApiRunEvent(
                run_id="level-test-run",
                message=f"This is a {level.value} level message",
                timestamp="1641046800000",
                level=level,
                step_key="test_step" if level != RunEventLevel.INFO else None,
                event_type="MessageEvent",
                error=None,
            )
            for level in RunEventLevel
        ]

        log_list = RunEventList(items=events, total=len(events))

        with fixed_timezone("UTC"):
            result = format_logs_table(log_list, "level-test-run")

        snapshot.assert_match(result)


class TestLogDataProcessing:
    """Test processing of log data structures.

    This class tests the data model creation and processing logic
    for run events and error information.
    """

    def test_error_info_stack_trace_formatting(self):
        """Test DgApiErrorInfo stack trace string conversion."""
        error = DgApiErrorInfo(
            message="Test error message",
            className="TestError",
            stack=[
                '  File "/app/test.py", line 10, in test_function\n    raise TestError("Something went wrong")\n',
                '  File "/app/main.py", line 5, in main\n    test_function()\n',
            ],
            cause=None,
        )

        stack_trace = error.get_stack_trace_string()
        expected = (
            '  File "/app/test.py", line 10, in test_function\n'
            '    raise TestError("Something went wrong")\n\n'
            '  File "/app/main.py", line 5, in main\n'
            "    test_function()\n"
        )

        assert stack_trace == expected

    def test_error_info_empty_stack(self):
        """Test DgApiErrorInfo with empty stack."""
        error = DgApiErrorInfo(
            message="Error with no stack",
            className="EmptyStackError",
            stack=None,
            cause=None,
        )

        assert error.get_stack_trace_string() == ""

        # Test with empty list too
        error_empty_list = DgApiErrorInfo(
            message="Error with empty stack list",
            className="EmptyStackError",
            stack=[],
            cause=None,
        )

        assert error_empty_list.get_stack_trace_string() == ""

    def test_run_event_creation_with_all_levels(self, snapshot):
        """Test creating run events with all possible log levels."""
        events = [
            DgApiRunEvent(
                run_id=f"level-{level.value.lower()}-run",
                message=f"Message at {level.value} level",
                timestamp="1641046800000",
                level=level,
                step_key=f"step_{level.value.lower()}" if level != RunEventLevel.INFO else None,
                event_type="MessageEvent",
                error=None,
            )
            for level in RunEventLevel
        ]

        log_list = RunEventList(items=events, total=len(events))

        # Test JSON serialization works correctly for all levels
        result = log_list.model_dump_json(indent=2)
        import json

        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_run_event_list_pagination_handling(self):
        """Test RunEventList pagination properties."""
        events = [
            DgApiRunEvent(
                run_id=f"pagination-run-{i}",
                message=f"Message {i}",
                timestamp=f"164104680{i}000",
                level=RunEventLevel.INFO,
                step_key=None,
                event_type="MessageEvent",
                error=None,
            )
            for i in range(3)
        ]

        # Test with pagination info
        log_list = RunEventList(
            items=events,
            total=10,  # Total could be different from items length (pagination)
            cursor="next_page_cursor",
            has_more=True,
        )

        assert len(log_list.items) == 3
        assert log_list.total == 10
        assert log_list.cursor == "next_page_cursor"
        assert log_list.has_more is True

    def test_nested_error_cause_chain(self):
        """Test deeply nested error cause chains."""
        # Create a 3-level deep error chain
        root_error = DgApiErrorInfo(
            message="IOError: File not found",
            className="IOError",
            stack=[
                "  File \"/app/io.py\", line 15, in read_file\n    with open(filename, 'r') as f:\n"
            ],
            cause=None,
        )

        middle_error = DgApiErrorInfo(
            message="ProcessingError: Failed to process file",
            className="ProcessingError",
            stack=[
                '  File "/app/processor.py", line 25, in process\n    content = read_file(filename)\n'
            ],
            cause=root_error,
        )

        top_error = DgApiErrorInfo(
            message="PipelineError: Pipeline execution failed",
            className="PipelineError",
            stack=[
                '  File "/app/pipeline.py", line 50, in run\n    result = process(input_file)\n'
            ],
            cause=middle_error,
        )

        # Verify the chain is properly constructed
        assert top_error.cause is middle_error
        assert middle_error.cause is root_error
        assert root_error.cause is None

        # Test stack trace formatting works for all levels
        assert "Pipeline execution failed" in top_error.message
        assert top_error.cause is not None
        assert "Failed to process file" in top_error.cause.message
        assert top_error.cause.cause is not None
        assert "File not found" in top_error.cause.cause.message

    def test_run_event_with_no_optional_fields(self):
        """Test DgApiRunEvent creation with minimal required fields."""
        event = DgApiRunEvent(
            run_id="minimal-run",
            message="Minimal event",
            timestamp="1641046800000",
            level=RunEventLevel.INFO,
            # All optional fields are None by default
        )

        assert event.run_id == "minimal-run"
        assert event.message == "Minimal event"
        assert event.timestamp == "1641046800000"
        assert event.level == RunEventLevel.INFO
        assert event.step_key is None
        assert event.event_type is None
        assert event.error is None

    def test_timestamp_handling_edge_cases(self, snapshot):
        """Test timestamp formatting with edge case values."""
        from dagster_shared.utils.timing import fixed_timezone

        # Test various timestamp formats
        events = [
            DgApiRunEvent(
                run_id="timestamp-test",
                message="Epoch timestamp",
                timestamp="0",  # Unix epoch
                level=RunEventLevel.INFO,
                step_key=None,
                event_type="MessageEvent",
                error=None,
            ),
            DgApiRunEvent(
                run_id="timestamp-test",
                message="Large timestamp",
                timestamp="32503680000000",  # Year 3000
                level=RunEventLevel.DEBUG,
                step_key=None,
                event_type="MessageEvent",
                error=None,
            ),
        ]

        log_list = RunEventList(items=events, total=len(events))

        with fixed_timezone("UTC"):
            result = format_logs_table(log_list, "timestamp-test")

        snapshot.assert_match(result)
