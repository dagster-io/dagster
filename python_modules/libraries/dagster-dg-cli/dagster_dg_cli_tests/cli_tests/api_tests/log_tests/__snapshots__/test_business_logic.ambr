# serializer version: 1
# name: TestFormatLogs.test_format_empty_logs_json_output[0]
  dict({
    'count': 0,
    'cursor': None,
    'hasMore': False,
    'logs': list([
    ]),
    'run_id': 'empty-run-67890',
  })
# ---
# name: TestFormatLogs.test_format_empty_logs_table_output[0]
  'No logs found for run empty-run-67890'
# ---
# name: TestFormatLogs.test_format_logs_all_levels[0]
  '''
  Logs for run level-test-run:
  
  TIMESTAMP            LEVEL    STEP_KEY                  MESSAGE
  --------------------------------------------------------------------------------
  2022-01-01 14:20:00  CRITICAL test_step                 This is a CRITICAL level message
  2022-01-01 14:20:00  ERROR    test_step                 This is a ERROR level message
  2022-01-01 14:20:00  WARNING  test_step                 This is a WARNING level message
  2022-01-01 14:20:00  INFO                               This is a INFO level message
  2022-01-01 14:20:00  DEBUG    test_step                 This is a DEBUG level message
  
  Total log entries: 5
  '''
# ---
# name: TestFormatLogs.test_format_logs_json_output[0]
  dict({
    'count': 5,
    'cursor': None,
    'hasMore': False,
    'logs': list([
      dict({
        'error': None,
        'eventType': 'RunStartEvent',
        'level': 'INFO',
        'message': 'Starting execution of run for "test_pipeline".',
        'runId': 'test-run-12345',
        'stepKey': None,
        'timestamp': '1641046800000',
      }),
      dict({
        'error': None,
        'eventType': 'ExecutionStepStartEvent',
        'level': 'DEBUG',
        'message': 'Started execution of step "process_data".',
        'runId': 'test-run-12345',
        'stepKey': 'process_data',
        'timestamp': '1641046805000',
      }),
      dict({
        'error': None,
        'eventType': 'MessageEvent',
        'level': 'DEBUG',
        'message': 'Loading input from path: /tmp/input.json',
        'runId': 'test-run-12345',
        'stepKey': 'process_data',
        'timestamp': '1641046810000',
      }),
      dict({
        'error': dict({
          'cause': None,
          'className': 'ValueError',
          'message': '''
            ValueError: Invalid input data format
  
          ''',
          'stack': list([
            '''
                File "/app/pipeline.py", line 42, in process_data
                  data = json.loads(input_str)
  
            ''',
            '''
                File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
                  return _default_decoder.decode(s)
  
            ''',
            '''
                File "/usr/lib/python3.12/json/decoder.py", line 337, in decode
                  obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  
            ''',
          ]),
          'stackTrace': '''
              File "/app/pipeline.py", line 42, in process_data
                data = json.loads(input_str)
            
              File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
                return _default_decoder.decode(s)
            
              File "/usr/lib/python3.12/json/decoder.py", line 337, in decode
                obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  
          ''',
        }),
        'eventType': 'ExecutionStepFailureEvent',
        'level': 'ERROR',
        'message': 'Execution of step "process_data" failed.',
        'runId': 'test-run-12345',
        'stepKey': 'process_data',
        'timestamp': '1641046815000',
      }),
      dict({
        'error': None,
        'eventType': 'RunFailureEvent',
        'level': 'ERROR',
        'message': 'Execution of run for "test_pipeline" failed. Steps failed: [\'process_data\'].',
        'runId': 'test-run-12345',
        'stepKey': None,
        'timestamp': '1641046820000',
      }),
    ]),
    'run_id': 'test-run-12345',
  })
# ---
# name: TestFormatLogs.test_format_logs_table_output[0]
  '''
  Logs for run test-run-12345:
  
  TIMESTAMP            LEVEL    STEP_KEY                  MESSAGE
  --------------------------------------------------------------------------------
  2022-01-01 14:20:00  INFO                               Starting execution of run for "test_pipeline".
  2022-01-01 14:20:05  DEBUG    process_data              Started execution of step "process_data".
  2022-01-01 14:20:10  DEBUG    process_data              Loading input from path: /tmp/input.json
  2022-01-01 14:20:15  ERROR    process_data              Execution of step "process_data" failed.
  
  Stack Trace:
      File "/app/pipeline.py", line 42, in process_data
        data = json.loads(input_str)
      File "/usr/lib/python3.12/json/__init__.py", line 346, in loads
        return _default_decoder.decode(s)
      File "/usr/lib/python3.12/json/decoder.py", line 337, in decode
        obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  
  2022-01-01 14:20:20  ERROR                              Execution of run for "test_pipeline" failed. Steps failed: ['process_data'].
  
  Total log entries: 5
  '''
# ---
# name: TestFormatLogs.test_format_logs_with_long_step_key[0]
  '''
  Logs for run truncation-test-run:
  
  TIMESTAMP            LEVEL    STEP_KEY                  MESSAGE
  --------------------------------------------------------------------------------
  2022-01-01 14:20:30  INFO     very_long_step_name_that  Processing large dataset chunk.
  
  Total log entries: 1
  '''
# ---
# name: TestFormatLogs.test_format_logs_with_nested_errors[0]
  '''
  Logs for run nested-error-run:
  
  TIMESTAMP            LEVEL    STEP_KEY                  MESSAGE
  --------------------------------------------------------------------------------
  2022-01-01 14:20:25  ERROR    database_query            Database connection failed with retry exhausted.
  
  Stack Trace:
      File "/app/database.py", line 25, in execute_query
        return self._execute_with_retry(query)
      File "/app/database.py", line 45, in _execute_with_retry
        raise RetryRequestedFromPolicy(f"Exceeded max_retries of {max_retries}")
  
  
  Total log entries: 1
  '''
# ---
# name: TestFormatLogs.test_format_logs_with_nested_errors_json[0]
  dict({
    'count': 1,
    'cursor': None,
    'hasMore': False,
    'logs': list([
      dict({
        'error': dict({
          'cause': dict({
            'cause': None,
            'className': 'ConnectionError',
            'message': '''
              ConnectionError: [Errno 111] Connection refused
  
            ''',
            'stack': list([
              '''
                  File "/app/database.py", line 35, in _execute_with_retry
                    result = self.connection.execute(query)
  
              ''',
              '''
                  File "/usr/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
                    return connection.execute(statement, parameters)
  
              ''',
              '''
                  File "/usr/lib/python3.12/site-packages/psycopg2/__init__.py", line 122, in connect
                    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
  
              ''',
            ]),
            'stackTrace': '''
                File "/app/database.py", line 35, in _execute_with_retry
                  result = self.connection.execute(query)
              
                File "/usr/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
                  return connection.execute(statement, parameters)
              
                File "/usr/lib/python3.12/site-packages/psycopg2/__init__.py", line 122, in connect
                  conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
  
            ''',
          }),
          'className': 'RetryRequestedFromPolicy',
          'message': '''
            RetryRequestedFromPolicy: Exceeded max_retries of 3
  
          ''',
          'stack': list([
            '''
                File "/app/database.py", line 25, in execute_query
                  return self._execute_with_retry(query)
  
            ''',
            '''
                File "/app/database.py", line 45, in _execute_with_retry
                  raise RetryRequestedFromPolicy(f"Exceeded max_retries of {max_retries}")
  
            ''',
          ]),
          'stackTrace': '''
              File "/app/database.py", line 25, in execute_query
                return self._execute_with_retry(query)
            
              File "/app/database.py", line 45, in _execute_with_retry
                raise RetryRequestedFromPolicy(f"Exceeded max_retries of {max_retries}")
  
          ''',
        }),
        'eventType': 'ExecutionStepFailureEvent',
        'level': 'ERROR',
        'message': 'Database connection failed with retry exhausted.',
        'runId': 'nested-error-run',
        'stepKey': 'database_query',
        'timestamp': '1641046825000',
      }),
    ]),
    'run_id': 'nested-error-run',
  })
# ---
# name: TestFormatLogs.test_format_logs_with_pagination_info[0]
  '''
  Logs for run paginated-run:
  
  TIMESTAMP            LEVEL    STEP_KEY                  MESSAGE
  --------------------------------------------------------------------------------
  2022-01-01 14:20:00  INFO                               First log entry
  2022-01-01 14:20:05  DEBUG    step_1                    Second log entry
  
  Total log entries: 2
  Note: More logs available (use --limit to increase or --cursor to paginate)
  '''
# ---
# name: TestLogDataProcessing.test_run_event_creation_with_all_levels[0]
  dict({
    'cursor': None,
    'has_more': False,
    'items': list([
      dict({
        'error': None,
        'event_type': 'MessageEvent',
        'level': 'CRITICAL',
        'message': 'Message at CRITICAL level',
        'run_id': 'level-critical-run',
        'step_key': 'step_critical',
        'timestamp': '1641046800000',
      }),
      dict({
        'error': None,
        'event_type': 'MessageEvent',
        'level': 'ERROR',
        'message': 'Message at ERROR level',
        'run_id': 'level-error-run',
        'step_key': 'step_error',
        'timestamp': '1641046800000',
      }),
      dict({
        'error': None,
        'event_type': 'MessageEvent',
        'level': 'WARNING',
        'message': 'Message at WARNING level',
        'run_id': 'level-warning-run',
        'step_key': 'step_warning',
        'timestamp': '1641046800000',
      }),
      dict({
        'error': None,
        'event_type': 'MessageEvent',
        'level': 'INFO',
        'message': 'Message at INFO level',
        'run_id': 'level-info-run',
        'step_key': None,
        'timestamp': '1641046800000',
      }),
      dict({
        'error': None,
        'event_type': 'MessageEvent',
        'level': 'DEBUG',
        'message': 'Message at DEBUG level',
        'run_id': 'level-debug-run',
        'step_key': 'step_debug',
        'timestamp': '1641046800000',
      }),
    ]),
    'total': 5,
  })
# ---
# name: TestLogDataProcessing.test_timestamp_handling_edge_cases[0]
  '''
  Logs for run timestamp-test:
  
  TIMESTAMP            LEVEL    STEP_KEY                  MESSAGE
  --------------------------------------------------------------------------------
  1970-01-01 00:00:00  INFO                               Epoch timestamp
  3000-01-01 00:00:00  DEBUG                              Large timestamp
  
  Total log entries: 2
  '''
# ---
